Feature/dnd duration dropdown

lolostheman opened the pull request:
Added drop down menu for users to enable Do Not Disturb (DND) mode for a selectable duration, after which the app will automatically restore the original notification settings.

First had the logic in the dnd-util.ts. Got this working, but wanted to handle the dnd-toggle request logic in app/main/index.ts to better follow Zulip's contribuiting guide.

We listen for our LeftSidebarEvents() & dndButton EventListener() aka our renderer to send a message to our ipcMain.on() in app/main/index.ts. This is where we now handle the logic and send message back to the renderer with the new parameters.

Had some small UI issues, with modal overlays covering things. Was able to fix. Tested functionality with numerous test cases, logging results to console. Haven't found an edge case that breaks the code.

Feel free to comment on style or content within the dropdown.

This is my first open source contribution. Please bare with me. Thanks!


Fixes: (https://github.com/zulip/zulip-desktop/issues/561)


Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/user-attachments/assets/b87ad203-c8e0-4cb1-90ab-725e56fa1a0d)
![issue#561](https://github.com/user-attachments/assets/0b780340-1515-42f0-a156-2380369a5986)


**Platforms this PR was tested on:**

- [x] Windows
- [ ] macOS
- [ ] Linux (specify distro)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on app/renderer/css/main.css:
```diff
@@ -378,6 +378,27 @@ webview.focus {
+
+.dropdown div:hover {
+  background-color: rgb(68 68 68 / 100%);
+}
```
This is not well-scoped CSS; you don't want to apply CSS to entire rules like `div` for readability and fragility reasons.

lolostheman replied:
Handled this by applying style to specific class.

timabbott commented on app/renderer/js/main.ts:
```diff
@@ -1210,6 +1231,13 @@ window.addEventListener("load", async () => {
             >
           </div>
+          <div id="dnd-dropdown" class="dropdown hidden">
+            <div data-minutes="1">1 minute</div>
```
1 minute seems basically useless in practice -- is that just for testing?

lolostheman replied:
Yes, it was for testing. I have removed it in the commits.

timabbott commented:
@lolostheman thanks for the PR! Can you clean up the commit history to make it easier to review? The commits here seem to have a lot of churn where one commit is fixing up issues in the previous one. Especially code moves are best done in commits that don't make any changes other than moving the code and adjusting imports.

Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for advice on how to do a nicely reviewable PR.

andersk commented:
If the app exits (or the computer crashes, etc.) before the timer expires, this will make the temporary setting permanent. I don’t think that’s what a user would expect.

lolostheman committed:
typed_ipc: Add duration param to toggle-dnd RendererMessage as well as add new toggle-dnd-request RenderMessage.

Introduces an optional duration parameter to the toggle-dnd-request message, enabling the renderer to specify a time value for automatically reverting Do Not Disturb (DND) mode. The toggle-dnd-request event is now used to send this duration from the client to the main process. The toggle-dnd response has been updated to include the duration, allowing the renderer to reflect the timeout state accurately.

lolostheman committed:
tooltips: Add HTML structure and CSS for DND duration dropdown.

Introduces a hidden dropdown menu with preset duration options and accompanying styles. This prepares the UI for timed DND selection.

lolostheman committed:
renderer: Add event listener to show/hide DND duration dropdown.

Implements click and mouseleave behavior for the DND button and duration dropdown, allowing the user to select a duration visually. We also remove the dnd-util.js import, since we no longer handle the DNDtoggle() here and instead we send the toggle request to the ipcMain.

lolostheman committed:
main: Handle toggle-dnd-request and auto-revert DND in ipcMain.

When a duration is provided, sets a timer to automatically revert DND mode. Ensures only one timer is active at a time, and cleans up appropriately when DND is turned off early. We import dnd-util.js to handle the DNDUtil.toggle() in our timer logic.
Fixes #561.

lolostheman commented:
> If the app exits (or the computer crashes, etc.) before the timer expires, this will make the temporary setting permanent. I don’t think that’s what a user would expect.

Yes, this would indeed happen. if the app crashes before the setTimeout callback fires. I'll work on a fix. Probably by storing the duration in the config file, and then we can check if the duration has expired on startup.

lolostheman committed:
main: Revert DND on startup if expiration has passed.

Checks the dndExpiration config key on app startup. If the timestamp has passed, DND is toggled off. If still active, schedules a timeout to revert DND later.

alya commented:
@lolostheman Is this PR ready for the next round of review? It's helpful to post a comment explicitly noting this if so.

lolostheman commented:
> @lolostheman Is this PR ready for the next round of review? It's helpful to post a comment explicitly noting this if so.

@alya Yes, the PR is ready for another review.

lolostheman commented:
@alya are their any other steps I should take to try to push this PR to a reviewer? Or is it just a waiting game?

timabbott commented on app/renderer/js/main.ts:
```diff
@@ -1210,6 +1210,15 @@ window.addEventListener("load", async () => {
+            <div class="dnd-dropdown-item" data-minutes="180">3 hours</div>
+            <div class="dnd-dropdown-item" data-minutes="720">12 hours</div>
+            <div class="dnd-dropdown-item" data-minutes="forever">
+              Until I resume
```
These strings need to al be tagged for translation. Look at the adjacent strings for examples of how to do this properly.

timabbott commented:
@lolostheman sorry for the slow reply. In general, PRs that look like we can predictably merge them quickly tend to get faster reviews. Leaving stale testing code and having readily visible issues (like no translation tags) generally result in PRs being delayed until we have more time for them.

I don't personally work in this project much. I expect @shubham-padia will be reviewing this eventually as he's ramping up on the codebase. You're certainly welcome to work on another issue in the project if you'd like! I am hopeful that we'll be able to be more responsive to new-to-OSS PRs in the coming weeks than we have been historically.

lolostheman commented:
I noticed this repo wasn't very active before I started to even work on this issue. So that being said I was well aware of the potential slow PR reviews. Thanks for taking the time to look at my PR once again. I rebased and added the translation tags. Let me know if there is any thing else. Take your time at reviewing, no rush. @timabbott
